core/client/hammer/hammer.go

```diff
@@ -76,125 +96,187 @@ func New(ctx context.Context, dial DialFunc, callOptions CallOptions,
+			userIDs := make([]string, 0, batch)
+			for j := 0; j < batch; j++ {
+				userIDs = append(userIDs, fmt.Sprintf("user_%v", i*batch+j))
+			}
+			if err := rateLimiter.WaitN(cctx, batch); err != nil {
+				log.Printf("Duration %v elapsed, stopping request generation", duration)
 				return
-			case jobs <- i:
+			}
+			inflightReqs <- request{
+				UserIDs:  userIDs,
+				PageSize: batch,
 			}
 		}
 	}()
-	return jobs
+	return inflightReqs
 }
 
-// printStats prints stats repeatedly every period until outputs closes.
-func (h *Hammer) printStats(period time.Duration, outputs <-chan error) {
-	startTime := time.Now()
-	var totalRequests float64
-
-	// Print stats asyncronously until refresh closes.
-	refresh := time.NewTicker(period)
-	stop := make(chan struct{})
-	var wg sync.WaitGroup
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		for {
-			select {
-			case <-refresh.C:
-				rate := totalRequests / time.Since(startTime).Seconds()
-				log.Printf("QPS: %v, \tRequests: %v, \tWorkers: %v", rate, totalRequests, h.workers)
-			case <-stop:
-				rate := totalRequests / time.Since(startTime).Seconds()
-				log.Printf("QPS: %v, \tRequests: %v, \tWorkers: %v", rate, totalRequests, h.workers)
-				refresh.Stop()
-				return
-			}
-		}
-	}()
+type worker struct {
+	*Hammer
+	client *client.Client
+}
 
-	// Collect outputs until output closes.
-	for err := range outputs {
-		totalRequests++
+func (h *Hammer) newWorkers(n int) ([]worker, error) {
+	workers := make([]worker, 0, n)
+	for i := 0; i < n; i++ {
+		// Give each worker its own client.
+		client, err := client.NewFromConfig(h.ktCli, h.domain)
 		if err != nil {
-			log.Printf("writeOp(): %v", err)
+			return nil, err
 		}
+
+		workers = append(workers, worker{Hammer: h, client: client})
 	}
-	close(stop)
-	wg.Wait()
+	return workers, nil
 }
 
-func (h *Hammer) worker(ctx context.Context, jobs <-chan int, outputs chan<- error) {
-	client, err := client.NewFromConfig(h.ktCli, h.config)
-	if err != nil {
-		glog.Errorf("NewFromConfig(): %v", err)
-		return
+// batchWriteOp queues many user mutations, waits, and then verifies them all.
+func (w *worker) writeOp(ctx context.Context, req *request) error {
+	users := make([]*tpb.User, 0, len(req.UserIDs))
+	for _, userID := range req.UserIDs {
+		users = append(users, &tpb.User{
+			DomainId:       w.domain.DomainId,
+			AppId:          w.appID,
+			UserId:         userID,
+			PublicKeyData:  []byte("publickey"),
+			AuthorizedKeys: w.authorizedKeys,
+		})
+	}
+
+	mutations := make([]*entry.Mutation, 0, len(users))
+	for _, u := range users {
+		callOptions := w.callOptions(u.UserId)
+		cctx, cancel := context.WithTimeout(ctx, w.timeout)
+		defer cancel()
+		fmt.Print(".")
+		m, err := w.client.CreateMutation(cctx, u)
+		if err != nil {
+			fmt.Print("!")
+			return err
+		}
+		mutations = append(mutations, m)
+		if err := w.client.QueueMutation(cctx, m, w.signers, callOptions...); err != nil {
+			fmt.Print("!")
+			return err
+		}
+		fmt.Print("-")
```

Comment:
Can this `Print` and other similar ones be interleaved because of having multiple concurrent workers? If so, perhaps a better approach would be to construct a string first and print it with a single call?

Reply:
These prints can be interleaved.  That said, I'm still searching for a better way to output status from many workers to the screen.